added the ability to filter columns with array values

DevelopmentByDavid opened the pull request:
Allows the ability to filter based on array values in a single cell.  Addresses Issue #546.
Please let me know if there is anything additional I should do or change.  This is my first time creating a PR and contributing to open source! :smile:  

Example pictures below.

![image](https://user-images.githubusercontent.com/44346401/56004036-19343f80-5c7e-11e9-9db4-3c80361e2250.png)
![image](https://user-images.githubusercontent.com/44346401/56004046-22251100-5c7e-11e9-8cd4-1105551ab58a.png)

gregnb commented:
hey @DevelopmentByDavid just returning from a long vacation. Need some time to review. This PR from a quick glance is also asking to make a big change to the way the package accepts data by allowing: strings, numbers, and arrays? 

What would happen if a user places an array but does not use a customBodyRender? how does that display?

I think it's a good idea, but I need some time to think about this major change of allows Array data

DevelopmentByDavid commented:
Hi @gregnb  welcome back! Hope your vacation was good.  

I believe that the way in which I modified the code only effects *filtering* arrays of values in a single cell.  Shown below is the filtering done without the customBodyRender. 

If a person has an array of values in all cells of a column, then they can specify all possible filterOptions (from the docs).  However, my problem was that I had no way to filter that column (the one with arrays as values) appropriately.  I wanted the row to display if it contained any of the _tags_ in that particular cell.

Let me know if I need to add better comments or do something else! Happy and willing to help if I can.
![image](https://user-images.githubusercontent.com/44346401/56082545-dfbf1980-5dce-11e9-83ab-a11118542bcd.png)

DevelopmentByDavid commented:
Hey @gregnb any update on this PR? I'm able to update this more if needed.

gregnb commented:
Hey @DevelopmentByDavid can you add some unit tests for this enhancement? Test a case with arrays and without

DevelopmentByDavid commented:
Hey @gregnb quick question before I code up the tests.  

Would you rather me modify the data here (link below), which results in me having to modify other tests, or would you rather me build the 2 tests you mentioned with their own set of data?

I think if I modify the original data below it makes sure the changes integrate well with everything else. However, there are a decent number of tests where I have to change what the expected value is.    

Let me know what you prefer/think is better and I will do it.  Thanks!

 https://github.com/gregnb/mui-datatables/blob/b972ad44d5d57c77668017db65eb19d07e788e3c/test/MUIDataTable.test.js#L25-L67,

gregnb commented:
@DevelopmentByDavid I would prefer new tests that way it will be easy to track the new enhancement

DevelopmentByDavid commented:
Hey @gregnb 

So, I changed a few things from my original PR code-wise to be more consistent with how the table already works.  I also added one more enhancement so that the table _just works_ with arrays of data.  Here's the breakdown of everything I added.

- New example  in `array-value-columns/index`
- Filtering based on arrays of data is now more consistent with how the table already works (I'll clarify below)
- `filterData` is extracted from arrays properly, now you don't need to specify `filterOptions` if there are arrays of data in the table -- nice for arrays _just working_.
- Wrote 4 new test: array filtering, array filter option extraction, non-array filtering, non-array filter option extraction

From bullet 2, I realized that all data in the `filterList` is logically AND'd  and that with array data, I was essentially doing logical OR's -- array data is now logically AND'd.  

Please let me know if I need to explain anything further, add more tests, cleanup/modify code, etc.  Thanks!  :smile:

gregnb commented:
Great! Thanks for making the changes. Merged!
